src/QsCompiler/DataStructures/Diagnostics.fs

```diff
@@ -7,6 +7,9 @@ open System
 open System.Collections.Generic
 
 type ErrorCode =
+    | TypeMismatch = 1
```

Comment:
I wonder if errors starting at 1001 was a conscious decision to show all Q# compilation errors look similar (e.g. _error QS2003_). Would it make sense to start type errors at 4001?

Reply:
Yeah, the error codes I picked are unusual compared to the existing ones. I think there are two separate questions:

1. Should error codes start at 0001 or 1001?
2. Are there groups (first digit, first two digits, etc.) of error codes for different kinds of errors?

For question 1, I would rather format error codes with leading zeroes than require that all error codes start at 1001, to avoid error code inflation. For comparison, [FS0001](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/compiler-messages/fs0001) is the F# error code for type mismatch.

For question 2, I'm not sure that there is a consistent grouping of error codes by leading digits. For example, ExpectingIntExpr and ArgumentTypeMismatch are both type errors, but ExpectingIntExpr is QS5012 and ArgumentTypeMismatch is QS6210. ExpectingSemicolon and ExpectingIdentifier are both syntax errors, but ExpectingSemicolon is QS2003 and ExpectingIdentifier is QS6103.

Reply:
I agree with you on both answers. Just before closing this conversation, with this change if there is a TypeMismatch error would the user see QS0001? When the error is formatted to be shown to the user is it adding the leading zeroes?

Reply:
If there is a pattern to error code number groups, then I think it would help to have comments that describe each group, since I don't think that they are very clear from context. I think @bettinaheim might know what the groups are?

Reply:
> Just before closing this conversation, with this change if there is a TypeMismatch error would the user see QS0001? When the error is formatted to be shown to the user is it adding the leading zeroes?

Currently, it's formatted as QS1, but I can change it now to add leading zeroes.

Reply:
It's formatted with leading zeroes now (QS0001).